src/main/java/com/aws/greengrass/clientdevices/auth/util/WildcardTrie.java

```diff
@@ -0,0 +1,334 @@
+                    // check if # wildcard usage is valid
+                    if (i > 0 && subject.charAt(i - 1) == levelSeparatorChar) {
+                        current.isMQTTWildcard = true;
+                        current.matchAll = true;
+                        terminalLevel.isTerminalLevel = true;
+                    }
+                    return current;
+                }
+                return current.add(subject.substring(i + 1), true);
+            }
+            if (currentChar == singleLevelWildcardChar) {
+                current = current.add(sb.toString(), false);
+                current = current.children.get(MQTT_SINGLELEVEL_WILDCARD);
+                if (i == subjectLength - 1) {
+                    current.isTerminal = true;
+                    // check if '+' wildcard usage is valid
+                    // if it's used at the last level
+                    if (i > 0 && subject.charAt(i - 1) == levelSeparatorChar) {
+                        current.isMQTTWildcard = true;
+                    }
+                    return current;
+                }
+                // check if '+' wildcard usage is valid
+                // if it's used in middle levels
+                if (i > 0 && subject.charAt(i - 1) == levelSeparatorChar
+                        && subject.charAt(i + 1) == levelSeparatorChar) {
+                    current.isMQTTWildcard = true;
+                }
+                return current.add(subject.substring(i + 1), true);
+            }
+            if (currentChar == escapeChar) {
+                char actualChar = getActualChar(subject.substring(i));
+                if (actualChar != nullChar) {
+                    sb.append(actualChar);
+                    i = i + 3;
+                    continue;
+                }
+            }
+            sb.append(currentChar);
+        }
+        // Handle non-wildcard value
+        current = current.children.get(sb.toString());
+        current.isTerminal |= isTerminal;
+        return current;
+    }
+
+    /**
+     * The method tries to parse the given string using escape sequence ${c} (where c is a character to be escaped)
+     * and returns the character c if the pattern is matched. In any other scenario it returns null character ('\0')
+     *
+     * @param str string provided to get
+     */
+    static char getActualChar(String str) {
+        if (str.length() < 4) {
+            return nullChar;
+        }
+        // Match the escape format ${c}
+        if (str.charAt(0) == escapeChar && str.charAt(1) == '{' && str.charAt(3) == '}') {
+            return str.charAt(2);
+        }
+        return nullChar;
+    }
+
+    /**
+     * Match given string to the corresponding allowed resources trie. MQTT wildcards are not processed.
+     *
+     * @param str string to match.
+     */
+    @SuppressWarnings({"PMD.UselessParentheses", "PMD.CollapsibleIfStatements"})
+    public boolean matchesStandard(String str) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 18. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.